nav2_planner/src/planner_server.cpp

```diff
@@ -656,16 +666,58 @@ void PlannerServer::isPathValid(
     std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
     unsigned int mx = 0;
     unsigned int my = 0;
+    nav2_costmap_2d::Footprint footprint;
+
+    if (is_path_valid_check_footprint_) {
+      footprint = costmap_ros_->getRobotFootprint();
+    }
+
+    // Visualize the first polygon on the path
+    unsigned int cost = nav2_costmap_2d::FREE_SPACE;
     for (unsigned int i = closest_point_index; i < request->path.poses.size(); ++i) {
-      costmap_->worldToMap(
-        request->path.poses[i].pose.position.x,
-        request->path.poses[i].pose.position.y, mx, my);
-      unsigned int cost = costmap_->getCost(mx, my);
+      auto & position = request->path.poses[i].pose.position;
+      auto theta = tf2::getYaw(request->path.poses[i].pose.orientation);
+      if (is_path_valid_check_footprint_) {
+        costmap_ros_->getRobotFootprint();
+        cost = static_cast<unsigned int>(collision_checker_->footprintCostAtPose(
+            position.x, position.y, theta, footprint));
+      } else {
+        costmap_->worldToMap(
+          position.x,
+          position.y, mx, my);
+        cost = costmap_->getCost(mx, my);
+      }
 
-      if (cost == nav2_costmap_2d::LETHAL_OBSTACLE ||
-        cost == nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE)
+      if (!is_path_valid_check_footprint_ &&
+        (cost == nav2_costmap_2d::LETHAL_OBSTACLE ||
+        cost == nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE))
       {
         response->is_valid = false;
+        break;
+      } else if (cost == nav2_costmap_2d::LETHAL_OBSTACLE) {
+        // Vizualize the polygon that caused the failure
+        double cos_th = cos(theta);
+        double sin_th = sin(theta);
+        nav2_costmap_2d::Footprint oriented_footprint;
+        for (unsigned int j = 0; j < footprint.size(); ++j) {
+          geometry_msgs::msg::Point new_pt;
+          new_pt.x = position.x + (footprint[j].x * cos_th - footprint[j].y * sin_th);
+          new_pt.y = position.y + (footprint[j].x * sin_th + footprint[j].y * cos_th);
+          oriented_footprint.push_back(new_pt);
+        }
+
+        geometry_msgs::msg::PolygonStamped stamped_footprint_;
+        for (const auto point : oriented_footprint) {
+          geometry_msgs::msg::Point32 point_msg;
+          point_msg.x = point.x;
+          point_msg.y = point.y;
+          stamped_footprint_.polygon.points.push_back(point_msg);
+        }
+        stamped_footprint_.header.frame_id = "map";
+        failed_footprint_->publish(stamped_footprint_);
```

Comment:
Do you intend to keep the publisher or is that just a debugging artifact?

Reply:
I used it to verify my work :). I think it would be useful to keep around. However, I would clean it up and only pub on sub.

Reply:
What utility would this have beyond testing your code as working? I'm not sure how someone would use that in some workflow that would be useful, but maybe I'm not being imaginative enough

Reply:
Removed :)